### PR TITLE
KOGITO-2674: [DMN Designer] Multiple DRDs support - Context menu component

### PR DESCRIPTION
**Pull request by vpellegrino**
*Please refer to:* https://issues.redhat.com/browse/KOGITO-2674

*Task description:* Need a context menu for DRD actions

*Proposed solution:* Providing a generic context menu widget

[There](https://gist.github.com/vpellegrino/c388370bc58515ba741c2bf91f5832fa) you can find an utilization example of this widget.
The result is the follow:
context_menu_demo

### CODE REVIEW

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/contextmenu/ContextMenu.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+import elemental2.dom.HTMLElement;
+import org.jboss.errai.ui.client.local.api.elemental2.IsElement;
+import org.kie.workbench.common.dmn.client.widgets.grid.controls.list.HasListSelectorControl.ListSelectorHeaderItem;
+import org.kie.workbench.common.dmn.client.widgets.grid.controls.list.HasListSelectorControl.ListSelectorItem;
+import org.kie.workbench.common.dmn.client.widgets.grid.controls.list.HasListSelectorControl.ListSelectorTextItem;
+import org.uberfire.client.mvp.UberElemental;
+import org.uberfire.mvp.Command;
+
+public class ContextMenu {
```
Comment (karreiro) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/contextmenu/ContextMenu.java:33]: I do believe we could keep this component as a `@ApplicationScoped`, given the nature of its UI and API. Wdyt?

**Comment by vpellegrino** [comment]
@karreiro I updated PR including your suggestions. Thanks for the feedback!

**Comment by vpellegrino** [comment]
Jenkins please retest this

**Comment by vpellegrino** [comment]
> SonarCloud Quality Gate failed.
> 
> [Bug](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=BUG)
> [Vulnerability](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=VULNERABILITY) (and [Security Hotspot](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=SECURITY_HOTSPOT) to review)
> [Code Smell](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=CODE_SMELL) [1 Code Smell](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&metric=new_coverage&view=list) [83.1% Coverage](https://sonarcloud.io/component_measures?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&metric=new_coverage&view=list)
> [](https://sonarcloud.io/component_measures?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&metric=new_duplicated_lines_density&view=list)
> 
> warning The version of Java (1.8.0_202) you have used to run this analysis is deprecated and we will stop accepting it from October 2020. Please update to at least Java 11.
> Read more [here](https://sonarcloud.io/documentation/upcoming/)

For the code smell, the `Span` element was already present in the constructor. I have only added a further parameter: the `Icon`. TBH, I would not change it, in order to be sure to not introduce regressions.

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by vpellegrino** [comment]
> SonarCloud Quality Gate failed.
> 
> [Bug](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=BUG)
> [Vulnerability](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=VULNERABILITY) (and [Security Hotspot](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=SECURITY_HOTSPOT) to review)
> [Code Smell](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=CODE_SMELL) [1 Code Smell](https://sonarcloud.io/project/issues?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&metric=new_coverage&view=list) [91.5% Coverage](https://sonarcloud.io/component_measures?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&metric=new_coverage&view=list)
> [](https://sonarcloud.io/component_measures?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=org.kie.workbench%3Akie-wb-common&pullRequest=3357&metric=new_duplicated_lines_density&view=list)
> 
> warning The version of Java (1.8.0_202) you have used to run this analysis is deprecated and we will stop accepting it from October 2020. Please update to at least Java 11.
> Read more [here](https://sonarcloud.io/documentation/upcoming/)

The reason quality gate is failing:
` 0.0% Condition Coverage on New Code (is less than 80%)`

Please let me know if there is any action pending from my side: coverage is now 91.5%

**Comment by vpellegrino** [comment]
> @vpellegrino I haven't checked the code yet and kogito-channels.
> 
> I checked on business central using a `DecisionNavigatorView` you shared. It works pretty well on chrome. However I got and NullPointer and freezing screen on firefox [1].
> 
> I am concerned its caused because I touched code manually to have entry point for showing your new menu. I had idea if we can not use experimental API (documented here https://github.com/kiegroup/kie-docs/pull/1131/files) that displays feature implemented as experimental just if business central is started with `-Dappformer.experimental.features=true`. It would eliminate any mistake done by me when touching code locally. Not saying it is doable but at least worth to assume as I expect we will iterate over this feature longer time.
> 
> [1]
> Screenshot from 2020-07-20 15-31-43

@jomarko thanks for your quick feedback.
It turns out that under firefox, `event.path` is not defined (https://www.thetopsites.net/article/53508096.shtml), so I have used `composedPath()` method instead. I tested it locally and works fine! I was hoping that GWT layer was already providing some polyfilling

**Comment by vpellegrino** [comment]
Test failure [org.kie.workbench.common.services.backend.compiler.ConcurrentBuildTest.buildFourProjectsInFourThread](https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/KIE/job/master/job/pullrequest/job/new-PR-kie-wb-common-master.pullrequests/90/testReport/junit/org.kie.workbench.common.services.backend.compiler/ConcurrentBuildTest/buildFourProjectsInFourThread/) seems to be not related to this PR

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/contextmenu/ContextMenu.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    @PostConstruct
+    public void init() {
+        view.init(this);
+    }
+
+    public void show() { view.show(); }
+
+    public void show(Consumer<ContextMenu> contextMenuConsumer) {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/contextmenu/ContextMenu.java:55]: Could we please javadoc the consumer? From implementation I have some idea what it does but I think it is worth to write in few words.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/contextmenu/ContextMenu.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        this.resetMenuItems();
+        contextMenuConsumer.accept(this);
+        this.show();
+    }
+
+    public void hide() { view.hide(); }
+
+    public HTMLElement getElement() {
+        return view.getElement();
+    }
```
[the remark this answers is not in the document]
Comment (vpellegrino) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/contextmenu/ContextMenu.java:65]: Yes, it is used by the component that instantiates the ContextMenu, in order to add it to the DOM.

For example:
```
        final HTMLElement contextMenuElement = drdContextMenu.getElement();
        DomGlobal.document.body.appendChild(contextMenuElement);
```

**Comment by vpellegrino** [comment]
@kiegroup/gatekeepers can you please merge?

**Pull request merged**